### PR TITLE
feat(gatsby-plugin-gatsby-cloud): Rewrite in TS

### PR DESCRIPTION
**Pull request by leithonenglish**
## Description

- Re-write for the `gatsby-plugin-gatsby-cloud` package to use Typescript
- Add functionality to remove tooltip `using display none` by listening to the `transitionend event`

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -27,8 +27,8 @@ packages/gatsby-admin/public
 packages/gatsby/gatsby-admin-public
 packages/gatsby-codemods/transforms
 packages/gatsby-source-graphql/batching
-packages/gatsby-plugin-gatsby-cloud/components
-packages/gatsby-plugin-gatsby-cloud/utils
+packages/gatsby-plugin-gatsby-cloud/**/*.js
+!packages/gatsby-plugin-gatsby-cloud/src/**/*.js
```
Comment (pvorozhe) [inline .eslintignore:31]: why is this updated?

File: packages/gatsby-plugin-gatsby-cloud/package.json [hunk 2]
```diff
@@ -51,11 +51,11 @@
   },
   "sideEffects": false,
   "scripts": {
-    "build": "babel src --out-dir . --ignore \"**/__tests__\" && npm run clean && npm run copy-type-declarations",
+    "build": "babel src --out-dir . --ignore \"**/__tests__\" --extensions \".tsx,.ts,.js\" --source-maps && npm run clean && npm run copy-type-declarations",
```
Comment (LekoArts) [inline packages/gatsby-plugin-gatsby-cloud/package.json:54]: While you're at it you can also change the `--out-dir` to `dist` and then you don't need to edit the `.eslintignore` and it's more organized like in the other packages

Comment (TylerBarnes) [reply packages/gatsby-plugin-gatsby-cloud/package.json:54]: @LekoArts that's awesome! Has that always worked with Gatsby?

**Review by TylerBarnes — APPROVED** [review]
Looks good 👍

**Review by TylerBarnes — CHANGES_REQUESTED** [review]
Changing to "request changes" for Lennarts request

File: packages/gatsby-plugin-gatsby-cloud/package.json (hunk 2 above)
Comment (leithonenglish) [reply packages/gatsby-plugin-gatsby-cloud/package.json:54]: @LekoArts great idea for the cleanup but the change to the build folder might break the imports for gatsbyjs.com [here](https://github.com/gatsby-inc/mansion/blob/master/cloud/gatsbyjs.com/src/dashboard/src/features/ContentLoader/index.tsx#L18). @veryspry suggested that we hold off until we can do it properly and do a version bump

Comment (LekoArts) [reply packages/gatsby-plugin-gatsby-cloud/package.json:54]: @TylerBarnes Yes, Gatsby always considers the `main` first. Other packages in our monorepo that are purely TS build their files into `dist` and point `main` to it.

@leithonenglish that's a bad practice on .com then 😅  The proper way to expose and use `components` would be to use the `exports` field on `package.json` then. https://nodejs.org/api/packages.html#package-entry-points

So could be something like:

```
  "main": "dist/index.js",
  "exports": {
    ".": "./dist/index.js",
    "./components": "./dist/components.js"
  },
```

That doesn't require a major bump.

Comment (LekoArts) [reply packages/gatsby-plugin-gatsby-cloud/package.json:54]: Another note here: Why were source maps activated? We normally don't emit those

**Comment by LekoArts** [comment]
If the task "Add functionality to remove tooltip using display none by listening to the transitionend event" needs to get out I'd consider moving it into it's own PR and merging that, because for the TypeScript portion I'd argue that it makes more sense to move everything over at once as the package itself is contained enough to allow this. I won't die on that hill but I'd rather have it fully moved over than in this mixed state where it'll eventually be forgotten.

**Pull request closed without merging**